### PR TITLE
Encapsulate WriteSiteNotFound error

### PR DESCRIPTION
**Pull request by sergeyfrolov**
### 1. What does this change do, exactly?
Encapsulates writing of error response when no served vhosts match the requested host.
If HTTP version is 2.0 or above, 421 status code is written, otherwise - 404. Error now is a bit more descriptive.

### 2. Please link to the relevant issues.
N/A

### 3. Which documentation changes (if any) need to be made because of this PR?
N/A

### 4. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I am willing to help maintain this change if there are issues with it later

### CODE REVIEW

File: caddyhttp/httpserver/server.go [hunk 1]
```diff
@@ -490,6 +490,20 @@ func DefaultErrorFunc(w http.ResponseWriter, r *http.Request, status int) {
 	WriteTextResponse(w, status, fmt.Sprintf("%d %s\n", status, http.StatusText(status)))
 }
 
+const errStatusMisdirectedRequest = 421 // RFC 7540, 9.1.2
+// WriteSiteNotFound writes appropriate error code to w, signaling that
```
Comment (mholt) [inline caddyhttp/httpserver/server.go]: For readability, put an empty line before this comment and after the `const`.

File: caddyhttp/httpserver/server.go [hunk 2]
```diff
@@ -490,6 +490,20 @@ func DefaultErrorFunc(w http.ResponseWriter, r *http.Request, status int) {
...
+// requested host is not served by Caddy on a given port.
+func WriteSiteNotFound(w http.ResponseWriter, r *http.Request) {
+	var status int
+	if r.ProtoMajor >= 2 {
+		// TODO: use http.StatusMisdirectedRequest when it gets defined
+		status = errStatusMisdirectedRequest
+	} else {
+		status = http.StatusNotFound
+	}
```
Comment (mholt) [inline caddyhttp/httpserver/server.go]: Can shorten this whole section to:

```go
status := http.StatusNotFound
if r.ProtoMajor >= 2 {
	// TODO: use http.StatusMisdirectedRequest when it gets defined
	status = errStatusMisdirectedRequest
}
```

File: caddyhttp/httpserver/server.go [hunk 3]
```diff
@@ -490,6 +490,20 @@ func DefaultErrorFunc(w http.ResponseWriter, r *http.Request, status int) {
 	WriteTextResponse(w, status, fmt.Sprintf("%d %s\n", status, http.StatusText(status)))
 }
 
+const errStatusMisdirectedRequest = 421 // RFC 7540, 9.1.2
```
Comment (mholt) [inline caddyhttp/httpserver/server.go]: This isn't technically an `error` value, so maybe renaming this const to `httpStatusMisdirectedRequest` would be more fitting.

File: caddyhttp/httpserver/server.go [hunk 4]
```diff
@@ -490,6 +490,20 @@ func DefaultErrorFunc(w http.ResponseWriter, r *http.Request, status int) {
...
+// requested host is not served by Caddy on a given port.
+func WriteSiteNotFound(w http.ResponseWriter, r *http.Request) {
+	var status int
+	if r.ProtoMajor >= 2 {
+		// TODO: use http.StatusMisdirectedRequest when it gets defined
+		status = errStatusMisdirectedRequest
+	} else {
+		status = http.StatusNotFound
+	}
+	WriteTextResponse(w, status, fmt.Sprintf("%d No site %s served at this port\n", status, r.Host))
```
Comment (mholt) [inline caddyhttp/httpserver/server.go]: For clarity, rephrase to `Site %s is not served on this interface` - if that sounds good to you?

File: caddyhttp/httpserver/server.go (hunk 3 above)
Comment (sergeyfrolov) [inline caddyhttp/httpserver/server.go]: Agreed.

File: caddyhttp/httpserver/server.go (hunk 4 above)
Comment (sergeyfrolov) [inline caddyhttp/httpserver/server.go]: That sounds better, assuming it is fine to call this an interface

**Comment by francislavoie** [comment]
The RFC mentions:
>This status code MUST NOT be generated by proxies.
I haven't really looked at the code flow, but this seems to be true, right? This code doesn't get hit if proxying?

**Comment by mholt** [comment]
@francislavoie Yeah, if we're proxying, the user would receive a proxied site, rather than this error message.

**Comment by mholt** [comment]
Welcome to the project, @sergeyfrolov - thanks for your first contribution!

**Comment by sergeyfrolov** [comment]
@mholt thank you! Glad to be part of the project!

**Pull request merged**